### PR TITLE
PR 2950 alternative implementation

### PR DESCRIPTION
**Pull request by ssoloff**
Please review.  It's very rough.  There's additional improvements that can be made, but I ran out of time.  The TODOs should be addressed before merging.

(I didn't add a TODO for this, but we may want to keep the code from `ConcurrentOddsCalculator` that measured the time to create the first clone and bypassed the second and subsequent clones if the clone time was too large [i.e. it ended up only ran one simulation at a time].)

As I said in the original PR, I only did minimal performance testing.  Re-running the benchmarks you've run previously would be appreciated.

Take your time.  I won't be able to look at this again for about 10 hours. :smile:

---

The key to maintaining performance was to

* Make #cpus `GameData` clones in `setGameData()` (this appears to only be called once per AI turn)
* Make #cpus simulation contexts in `setCalculateData()` (this is called once for each territory the AI evaluates)
* Initialize each simulation context in `runSimulation()` and reset it before returning it to the queue (this is called once per simulation run)

---

`Configuration` is kind of ugly, but the immutability was necessary to ensure values are correctly published across threads without resorting to making everything `volatile`.  Maybe not as part of your PR, but I expect this smell will eventually go away because we will probably end up doing something similar to what Dan did in his original Battle Calculator redesign.  Namely, getting rid of all the `setXXX()` methods on the calculator and replace them with a single configuration object.

---

There's also some unrelated changes here.  That was mostly cleanup to help me better understand what was going on in the battle calculator.

### CODE REVIEW

File: src/main/java/games/strategy/triplea/oddsCalculator/ta/IOddsCalculator.java [hunk 1]
```diff
@@ -12,34 +12,25 @@
  * Interface to ensure different implementations of the odds calculator all have the same public methods.
  */
 public interface IOddsCalculator {
-  void setGameData(final GameData data);
+  void setGameData(GameData data);
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/triplea/oddsCalculator/ta/IOddsCalculator.java]: Why is the final keyword being removed in this interface from the parameters of multiple methods?

**Comment by RoiEXLab** [comment]
Benchmark 1
Benchmark 2
Benchmark 3

So as you already mentioned, the "correct" implementation takes longer, however it seems the memory usage is slightly reduced on my system. But 200+ seconds additional time shouldn't occur normally...
Ideally we really want to no longer depend on any GameData objects for the calculations, I believe the key for a minimum footprint and maximum speed is really to avoid using a GameData object, but cheap container objects that simply hold units and their respective attack and defense values, of course we probably want to re-use the existing battle code for this, or make the battle code use potential new, easier to understand code.
But this is also a task for another PR.

File: src/main/java/games/strategy/triplea/oddsCalculator/ta/IOddsCalculator.java (hunk 1 above)
Comment (ssoloff) [reply src/main/java/games/strategy/triplea/oddsCalculator/ta/IOddsCalculator.java]: Applying `final` to parameters of non-static or non-default interface methods has no effect, so it's just effectively noise.  Similar (but for a different reason) to why we don't apply `public abstract` to non-static/non-default interface methods.

**Comment by ssoloff** [comment]
@RoiEXLab 

> It would have been nice, if you would've seperated the general refactoring from the functional changes to make it clearer what is related to what, but that's complaining on a very high level, especially because you were under time pressure ^^

Yeah, I didn't actually expect to get anywhere when I started playing around with the code.  And by that time, it was too late. 😞 

> But this should be something for another PR... The diff is getting way too big again.

> Maybe, but also in a different PR.

Agreed on both points.

**Comment by ssoloff** [comment]
> So as you already mentioned, the "correct" implementation takes longer, however it seems the memory usage is slightly reduced on my system. But 200+ seconds additional time shouldn't occur normally...

@RoiEXLab Thanks for running the benchmarks. :+1:

So, from the two benchmarks that used 20,000 iterations, we see a difference of

* 2.5 ms/iteration (current) vs. 3.0 ms/iteration (new) => 20% degradation
* 79 ms/iteration (current) vs. 86 ms/iteration (new) => 9% degradation

It doesn't sound like much, but maybe there's something obvious causing that slow down.  Things should be a bit less crazy tomorrow, and I should have time to dig deeper into the performance.  I'll take a look at the existing `ConcurrentOddsCalculator` to see if there's an additional optimization or two we may be missing that explains the degradation relative to `master`.

**Comment by ssoloff** [comment]
@RoiEXLab The latest commits simply address the TODOs I left in from yesterday (one TODO was changed to a FIXME regarding the thread safety of public methods).  In addition, I removed `CountUpAndDownLatch`, which is no longer used with the removal of `ConcurrentOddsCalculator` (keep that in mind if you don't merge this PR).

I went through the `ConcurrentOddsCalculator` code looking for anything that would provide a performance improvement over the parallel stream implementation in this thread.  The only theory I came up with is how each calculator run is executed.

In the `ConcurrentOddsCalculator` implementation, it divides the runs across N tasks that run in N threads.  For example, if 1000 runs are requested on a system with two processors, two tasks are created that each execute 500 runs.  Each task runs on one of two threads.

In the parallel stream implementation, it divides the runs across M tasks that run in N threads.  For example, if 1000 runs are requested on a system with two processors, 1000 tasks are created that each execute one run.  Each task runs on one of two threads.

So the 10-20% difference we're seeing could simply be from the overhead of context switching so many tasks onto the worker threads.  I'll try an experiment whereby I change the `calculate()` implementation so that each task handles multiple runs similar to the legacy implementation.

**Comment by RoiEXLab** [comment]
Thanks @ssoloff The diff looks good 👍

**Comment by RoiEXLab** [comment]
@ssoloff Do you still have any plans regarding this PR?
Or should I just merge the changes in and try to split up the tasks into even pieces?

**Comment by ssoloff** [comment]
@RoiEXLab I was going to try that split theory later today, but if you want to attempt it, go ahead and merge this PR. :smile:

**Comment by RoiEXLab** [comment]
@ssoloff I'll see if I get there in the next couple of hours. I'll merge the PR when I get to it, otherwise feel free to proceed.

**Comment by ssoloff** [comment]
@RoiEXLab I gave it a shot.  Unfortunately, the performance doesn't appear to change. 😞

**Comment by RoiEXLab** [comment]
@ssoloff Well at least that speaks for the Stream API, but still unfortunate.
I'm sure there's still room for improvement on the other internals though.

**Pull request merged**